src/models/orderbook.rs

```diff
@@ -0,0 +1,128 @@
+use crate::models::Order;
+use anyhow::Result;
+use ethcontract::web3::types::Address;
+use parking_lot::RwLock;
+use serde::Deserialize;
+use std::collections::HashMap;
+use std::sync::Arc;
+
+pub type OrderBookHashMap = HashMap<Address, HashMap<Address, Vec<Order>>>;
+
+#[derive(Clone, Deserialize)]
+pub struct OrderBook {
+    #[serde(with = "arc_rwlock_serde")]
+    pub orders: Arc<RwLock<OrderBookHashMap>>,
+}
+
+mod arc_rwlock_serde {
+    use parking_lot::RwLock;
+    use serde::de::Deserializer;
+    use serde::Deserialize;
+    use std::sync::Arc;
+
+    pub fn deserialize<'de, D, T>(d: D) -> Result<Arc<RwLock<T>>, D::Error>
+    where
+        D: Deserializer<'de>,
+        T: Deserialize<'de>,
+    {
+        Ok(Arc::new(RwLock::new(T::deserialize(d)?)))
+    }
+}
+
+impl OrderBook {
+    #[allow(dead_code)]
+    pub fn new() -> Self {
+        OrderBook {
+            orders: Arc::new(RwLock::new(HashMap::new())),
+        }
+    }
+    #[allow(dead_code)]
+    pub fn add_order(&mut self, order: Order) -> bool {
+        let mut current_orderbook = self.orders.write();
+        let layer_hash_map = current_orderbook.entry(order.sell_token).or_default();
+        let orders = layer_hash_map.entry(order.buy_token).or_default();
+        let search_result = orders.binary_search(&order.clone());
+        let pos = match search_result {
+            Err(e) => e,
+            Ok(_) => return false, // order is already existing
+        };
+        orders.insert(pos, order.clone());
+        true
+    }
+    #[allow(dead_code)]
+    pub fn get_orders_for_tokens(self, token_1: Address, token_2: Address) -> Result<Vec<Order>> {
+        let current_orderbook = self.orders.read();
+        let empty_hash_vec: Vec<Order> = Vec::new();
+        let empty_hash_map: HashMap<Address, Vec<Order>> = HashMap::new();
+        let new_hash_map = current_orderbook.get(&token_1).unwrap_or(&empty_hash_map);
+        Ok(new_hash_map
+            .get(&token_2)
+            .unwrap_or(&empty_hash_vec)
+            .clone())
```

Comment:
Just a cool little tidbit (although it effectively makes no difference in this case since empty `Vec`s don't perform any allocations):
```suggestion
            .cloned()
            .unwrap_or(empty_hash_vec))
```

This clones the inner value if `get(...)` returns `Some(vec)`. This way `empty_hash_vec` isn't created, cloned, then discarded.